### PR TITLE
LargeFileTaskUpload- High-level design changes 

### PR DESCRIPTION
**Pull request by nikithauc**
The design documents proposes changes to the LargeFileUploadTask.

Note - This contains only high-level design change plans, not a detailed description of the implementation changes.

### CODE REVIEW

**Comment by nikithauc** [comment]
Here is an example of a user handling stream upload - https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/320#issuecomment-753470522

File: design/large-file-upload-task-design.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+This document proposes high-level design modifications to the `LargeFileUploadTask` implementation. Reasons for change -
+
+-   Enhancement - Support Node.js Stream upload. Issue [#320](https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/320).
+-   Bug Fix - Support large file uploads to Outlook API. Issue [#359](https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/359).
+-   Enhancement- Support upload progress handler callback.  Issue [#305](https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/305).
+
+Outline of the current implementation -
+
+```
```
Comment (baywet) [inline design/large-file-upload-task-design.md:11]:
```suggestion
```TypeScript
```

File: design/large-file-upload-task-design.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    -   Add class `OutlookLargeFileUploadTask.ts` extending the `LargeFileUploadTask` similar to `OneDriveLargeFileUploadTask.ts`. This allows to handle Outlook API specific file upload customizations.
+    -   Currently the `upload` function in the `LargeFileUploadTask` returns only the response body. This can be changed to return the raw response received from the API which can filtered in the child class implementations.
+  
+###### 3. Support upload progress handler callback
+- Proposed changes -
+	- Add interface -> `interface Progress{
+    				progress(range: Range):void
+			   }`
+	- Add progressCallBack option to ->
+	  ```
```
Comment (baywet) [inline design/large-file-upload-task-design.md:86]:
```suggestion
	  ```TypeScript
```

File: design/large-file-upload-task-design.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+-   For browser support, take a dependency on [stream-browserify](https://www.npmjs.com/package/stream-browserify) - the stream module from node core, for browsers and make necessary updates to the rollup or bundling process.
+
+###### 2. Support large file uploads to Outlook API
+
+-   Bug in the current implemenation - `response.id !== undefined` this condition to mark the completion of an upload does not work for Outlook API since the final response does not contain a response body.
+-   An upload task should be marked as completed if the response status is a 201. [SDK-design document](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/FileUploadTask.md).
+-   The LargeFileUploadTask should allow uploads to OneDrive API, Outlook API and PrintDocument API.
+-   Proposed changes-
+    -   Add class `OutlookLargeFileUploadTask.ts` extending the `LargeFileUploadTask` similar to `OneDriveLargeFileUploadTask.ts`. This allows to handle Outlook API specific file upload customizations.
```
Comment (baywet) [inline design/large-file-upload-task-design.md:77]: In java I didn't need to go through specialization, having testing the response and it's body/headers allows to determine whether the upload is completed or whether a next chunk is expected. Do you feel we absolutely need a specialized class for the different service? Hopefully we don't end up with slightly different implementations for each service. It makes things "more complex" for consumers to have to think about "am I using outlook or onedrive?" which the promise of Microsoft Graph is to provide a unified API front (i.e. they don't necessarily know the distinction of services behind the scenes)

**Review by zengin — APPROVED** [review]

**Review by baywet — APPROVED** [review]

**Pull request merged**